### PR TITLE
Eg 355

### PR DESCRIPTION
**Pull request by woollett-r**
EG-355 Fix error when non-whitespace character immediately follows placeholder annotation

### CODE REVIEW

**Comment by cmrhmrc** [comment]
Looking at UIBuilder.fromInstructionGroup. You determine the first and second bps, then do the rest with a call to createRemainingItems. Couldnt this call do all of the Bps?

File: app/services/BulletPointBuilder.scala [hunk 1]
```diff
@@ -102,28 +106,42 @@ object BulletPointBuilder {
   def locateTextsAndMatchesContainingLeadingText(
```
Comment (cmrhmrc) [inline app/services/BulletPointBuilder.scala:106]: This fails the > 50 lines test. Its a difficult fn to follow, the boolean switch argument and the nested ifs compound situation. Splitting into two functions to remove boolean switch arg would produce two simpler fns

File: app/services/BulletPointBuilder.scala [hunk 2]
```diff
@@ -102,28 +106,42 @@ object BulletPointBuilder {
...
       noOfWordsToMatch: Int,
       inputTexts: List[String],
+      textsProcessed: Int,
       inputMatches: List[Match],
+      matchesProcessed: Int,
       outputTexts: Seq[String],
```
Comment (cmrhmrc) [inline app/services/BulletPointBuilder.scala:112]: There's a mixing of Lists and Seqs and subsequent toLists, would be better to stick with one or other and remove need for conversions

File: app/services/BulletPointBuilder.scala [hunk 3]
```diff
@@ -172,31 +192,42 @@ object BulletPointBuilder {
...
       texts: List[String],
+      textsProcessed: Int,
       matches: List[Match],
+      matchesProcessed: Int,
       items: List[String],
       matchText: Boolean,
       wordsProcessed: Int
   ): List[String] = {
 
     if (matchText) {
```
Comment (cmrhmrc) [inline app/services/BulletPointBuilder.scala:203]: like locateTextsAndMatchesContainingLeadingText() , this uses a boolean switch arg, to make the fn do two different things; work on texts or work on matches, would be clearer to create two fns

File: app/services/BulletPointBuilder.scala [hunk 4]
```diff
@@ -210,14 +241,35 @@ object BulletPointBuilder {
     *
     * @return - Returns a sub-sample of the input text
     */
-  def extractLeadingMatchedWords(noOfMatchedWords: Int, text: String): String = {
+  def extractLeadingMatchedWords(noOfMatchedWords: Int, texts: List[String], textsProcessed: Int, matches: List[Match], matchesProcessed: Int): String = {
+
+    val textElements: List[Match] = notSpaceRegex.findAllMatchIn(texts(textsProcessed)).toList
```
Comment (cmrhmrc) [inline app/services/BulletPointBuilder.scala:246]: think this is the same without the need to draw on regex and Matches
texts(textsProcessed).split(" +").toList

File: app/services/BulletPointBuilder.scala [hunk 5]
```diff
@@ -210,14 +241,35 @@ object BulletPointBuilder {
...
 
-    val matches: List[Match] = notSpaceRegex.findAllMatchIn(text).toList
+    if (matches.size == 0) {
```
Comment (cmrhmrc) [inline app/services/BulletPointBuilder.scala:248]: more usual to match a list with a case Nil =>

File: app/services/BulletPointBuilder.scala (hunk 2 above)
Comment (woollett-r) [reply app/services/BulletPointBuilder.scala:112]: Done

File: app/services/BulletPointBuilder.scala (hunk 1 above)
Comment (woollett-r) [reply app/services/BulletPointBuilder.scala:106]: Split function into two although this did mean relinquishing the tailrec annotation.

File: app/services/BulletPointBuilder.scala (hunk 3 above)
Comment (woollett-r) [reply app/services/BulletPointBuilder.scala]: I've split the function into two as requested.

File: app/services/BulletPointBuilder.scala (hunk 4 above)
Comment (woollett-r) [reply app/services/BulletPointBuilder.scala:255]: Running the test class the outputs of notSpaceRegex.findAllMatchIn(texts(textsProcessed)).toList and texts(textsProcessed).split(" +").toList are not always equal. In addition using a regex simplifies the extraction of the final element of the leading text. As such I recommend with keep the existing code.

File: app/services/BulletPointBuilder.scala [hunk 6]
```diff
@@ -247,4 +299,96 @@ object BulletPointBuilder {
   def getMatchText(m: Match): String = Option(m.group(1)).getOrElse(m.group(2))
 
   def wordsInString(text: String): Int = notSpaceRegex.findAllMatchIn(text).toList.size
```
[the remark this answers is not in the document]
Comment (woollett-r) [reply app/services/BulletPointBuilder.scala:310]: As previously, the two approaches do not always give the same result so changing to the text splitting approach would require a rework.

File: app/services/BulletPointBuilder.scala (hunk 5 above)
Comment (woollett-r) [reply app/services/BulletPointBuilder.scala:257]: I've rewritten the method use Scala match/case statements. I note, however, that I had to break the method into two to reduce the cyclomatic(?) complexity.

**Pull request merged**